torch/testing/_internal/common_distributed.py

```diff
@@ -515,14 +524,22 @@ def run_test(self, test_name: str, pipe) -> None:
         try:
             getattr(self, test_name)()
             # Close pipe after done with test.
-            pipe.close()
+            if signal_pipe is not None:
+                signal_pipe.send(None)
+            if event_listener_thread is not None:
+                event_listener_thread.join()
+            parent_pipe.close()
```

Comment:
Might be cleaner to put all the cleanup logic in a `finally` clause?

Reply:
How do you envision that working? We're doing slightly different things in the two cases (in particular in the except case we're sending the traceback over the `parent_pipe` and exiting). I don't see a clean way to reconcile them into a single codepath.

Reply:
I mean the signal_pipe and event_listener_thread conditions are the same in both cases right? That could go in a finally clause.

Reply:
Ok, I thought that `sys.exit` would immediately kill the process (and thus had to go last), while actually it's "just" raising an exception. Hence I guess we can do further cleanup after exiting (which is weird but ok).

BTW do you know why we're sending the traceback to the parent in case of an exception, instead of just logging it? It seems a more convoluted way of getting the same information in the end.